File: app/services/shipstation/v2/test_connection.rb
# frozen_string_literal: true

module Shipstation
  module V2
    # Verifies a company's ShipStation V2 API key by making an authenticated
    # read (GET /v2/carriers). Works for both production and TEST_ sandbox keys.
    # Returns { connected:, sandbox: } so the UI can confirm which environment
    # the key targets.
    class TestConnection < BaseService
      attr_reader :api_key

      def initialize(company_id)
        setting = IntegrationSetting.find_by(company_id: company_id)
        @api_key = setting&.v2_api_key
      end

      def call
        return result(false) if api_key.blank?

        response = HTTParty.get("#{SHIPSTATION_V2_API_BASE}/carriers", headers: headers)
        result(response.code == 200)
      rescue StandardError => e
        Rails.logger.error("[Shipstation::V2::TestConnection] #{e.class}: #{e.message}")
        result(false)
      end

    private

      def result(connected)
        { connected: connected, sandbox: api_key.to_s.start_with?("TEST_") }
      end
    end
  end
end

File: src/lib/shipstation/test-connection.ts
/**
 * Port of app/services/shipstation/test_connection.rb and
 * app/services/shipstation/v2/test_connection.rb.
 *
 * Verifies a company's stored credentials with an authenticated read-only
 * request. Both return "not connected" rather than throwing, because this is a
 * button in the config UI, not a decision the order path depends on.
 */

import { isSandboxKey } from "@/lib/integration-settings";

import {
  SHIPSTATION_V2_API_BASE,
  credentialsFor,
  hasV1Credentials,
  shipstationGet,
  v2Headers,
} from "./client";

export async function testV1Connection(companyId: bigint): Promise<boolean> {
  const credentials = await credentialsFor(companyId);
  if (!hasV1Credentials(credentials)) return false;

  try {
    const response = await shipstationGet("/carriers", credentials);
    return response.status === 200;
  } catch (error) {
    console.error(
      "[Shipstation::TestConnection]",
      error instanceof Error ? error.message : error,
    );
    return false;
  }
}

export interface V2ConnectionResult {
  connected: boolean;
  /** Which environment the stored key targets, so the UI can say so. */
  sandbox: boolean;
}

export async function testV2Connection(companyId: bigint): Promise<V2ConnectionResult> {
  const { v2ApiKey } = await credentialsFor(companyId);
  const sandbox = isSandboxKey(v2ApiKey);

  if (!v2ApiKey) return { connected: false, sandbox };

  try {
    const response = await fetch(`${SHIPSTATION_V2_API_BASE}/carriers`, {
      headers: v2Headers(v2ApiKey),
    });
    return { connected: response.status === 200, sandbox };
  } catch (error) {
    console.error(
      "[Shipstation::V2::TestConnection]",
      error instanceof Error ? error.message : error,
    );
    return { connected: false, sandbox };
  }
}
